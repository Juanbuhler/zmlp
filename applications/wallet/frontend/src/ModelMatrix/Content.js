import { useRouter } from 'next/router'
import useSWR from 'swr'

import { spacing } from '../Styles'

import ModelMatrixNavigation from './Navigation'
import ModelMatrixLayout from './Layout'

const ModelMatrixContent = () => {
  const {
    query: { projectId, modelId },
  } = useRouter()

  const { data: matrix } = useSWR(
    `/api/v1/projects/${projectId}/models/${modelId}/confusion_matrix/`,
  )

  return (
    <div
      css={{
        flex: 1,
        height: '100%',
        marginLeft: -spacing.spacious,
        marginRight: -spacing.spacious,
        marginBottom: -spacing.spacious,
        paddingTop: spacing.hairline,
        display: 'flex',
        flexDirection: 'column',
      }}
    >
      <ModelMatrixNavigation
        projectId={projectId}
        modelId={modelId}
        name={matrix.name}
      />

      <ModelMatrixLayout matrix={matrix} />
    </div>
  )
}

export default ModelMatrixContent
