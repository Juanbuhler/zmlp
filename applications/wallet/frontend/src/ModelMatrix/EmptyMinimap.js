import PropTypes from 'prop-types'

import { colors, constants, spacing } from '../Styles'

const EMPTY_MATRIX = [0, 1, 2, 3]

const ModelMatrixEmptyMinimap = ({ isMatrixApplicable }) => {
  return (
    <div
      css={{
        position: 'relative',
        border: isMatrixApplicable
          ? constants.borders.large.steel
          : constants.borders.large.smoke,
        padding: spacing.hairline,
        borderRadius: constants.borderRadius.small,
        display: 'grid',
        gridTemplate: `repeat(${EMPTY_MATRIX.length}, 1fr) / repeat(${EMPTY_MATRIX.length}, 1fr)`,
      }}
    >
      {EMPTY_MATRIX.map((row) => {
        return EMPTY_MATRIX.map((col) => {
          return (
            <div
              key={`${row}${col}`}
              css={{
                [`:not(:nth-of-type(4n))`]: {
                  borderRight: isMatrixApplicable
                    ? constants.borders.regular.coal
                    : constants.borders.regular.smoke,
                },
                [`:not(:nth-of-type(n + 13))`]: {
                  borderBottom: isMatrixApplicable
                    ? constants.borders.regular.coal
                    : constants.borders.regular.smoke,
                },
                paddingBottom: '100%',
                backgroundColor: isMatrixApplicable
                  ? colors.structure.smoke
                  : colors.structure.coal,
              }}
            />
          )
        })
      })}
    </div>
  )
}

ModelMatrixEmptyMinimap.propTypes = {
  isMatrixApplicable: PropTypes.bool.isRequired,
}

export default ModelMatrixEmptyMinimap
