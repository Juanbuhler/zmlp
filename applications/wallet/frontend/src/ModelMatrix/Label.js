import PropTypes from 'prop-types'
import { useState } from 'react'

import { colors, constants, spacing, zIndex } from '../Styles'

const ModelMatrixLabel = ({ cellDimension, label }) => {
  const [showLabel, setShowLabel] = useState(false)

  return (
    <div
      aria-label={`Prediction Label: ${label}`}
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      css={{ position: 'relative', cursor: 'default' }}
    >
      <div
        css={{
          minWidth: constants.borderWidths.large * 2,
          width: cellDimension,
          maxWidth: cellDimension,
          padding: cellDimension < spacing.normal * 2 ? 0 : spacing.normal,
          paddingTop: spacing.normal,
          paddingBottom: spacing.normal,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap',
          borderRight: constants.borders.regular.coal,
        }}
      >
        {label}
      </div>
      {showLabel && (
        <div
          css={{
            position: 'absolute',
            top: 0,
            left: 0,
            height: '100%',
            minWidth: cellDimension,
            padding: spacing.normal,
            backgroundColor: colors.structure.lead,
            zIndex: zIndex.layout.interactive,
            boxShadow: constants.boxShadows.default,
            whiteSpace: 'nowrap',
          }}
        >
          {label}
        </div>
      )}
    </div>
  )
}

ModelMatrixLabel.propTypes = {
  cellDimension: PropTypes.number.isRequired,
  label: PropTypes.string.isRequired,
}

export default ModelMatrixLabel
