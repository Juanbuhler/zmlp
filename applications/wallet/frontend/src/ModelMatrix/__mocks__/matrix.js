const matrix = {
  name: 'flowers',
  datasetId: '4b0b10a8-cec1-155c-b12f-ee2bc8787e06',
  labels: ['pink', 'red', 'yellow'],
  matrix: [
    [0, 0, 2],
    [1, 1, 0],
    [0, 0, 0],
  ],
  maxScore: 1,
  minScore: 0,
  moduleName: 'flower-types',
  overallAccuracy: 0.6,
  testSetOnly: true,
  isMatrixApplicable: true,
}

export default matrix
