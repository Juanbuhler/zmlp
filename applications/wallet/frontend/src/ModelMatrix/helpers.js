const GRADIENT = [
  [100, [0, 39, 107]],
  [63, [0, 149, 255]],
  [0, [255, 255, 255]],
]

const getClosestColors = ({ percent }) => {
  const index = GRADIENT.findIndex((c) => c[0] < percent)

  const leftBound = GRADIENT[index - 1] || GRADIENT[GRADIENT.length - 2]
  const rightBound = GRADIENT[index] || GRADIENT[GRADIENT.length - 1]

  const ratio = (percent - leftBound[0]) / (rightBound[0] - leftBound[0])

  return { leftColor: leftBound[1], rightColor: rightBound[1], ratio }
}

// http://jsfiddle.net/pwvz1gk9/
export const getColor = ({ percent: p }) => {
  const percent = Math.max(0, Math.min(p, 100))

  const { leftColor, rightColor, ratio } = getClosestColors({ percent })

  const i = ratio * 2 - 1
  const j = (i / 1 + 1) / 2
  const k = 1 - j

  const rgb = [
    Math.round(leftColor[0] * k + rightColor[0] * j),
    Math.round(leftColor[1] * k + rightColor[1] * j),
    Math.round(leftColor[2] * k + rightColor[2] * j),
  ]

  return `rgb(${rgb})`
}
