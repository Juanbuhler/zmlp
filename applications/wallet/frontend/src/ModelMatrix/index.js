import Head from 'next/head'

import SuspenseBoundary, { ROLES } from '../SuspenseBoundary'

import ModelMatrixContent from './Content'

const ModelMatrix = () => {
  return (
    <>
      <Head>
        <title>Confusion Matrix</title>
      </Head>

      <SuspenseBoundary role={ROLES.ML_Tools}>
        <ModelMatrixContent />
      </SuspenseBoundary>
    </>
  )
}

export default ModelMatrix
