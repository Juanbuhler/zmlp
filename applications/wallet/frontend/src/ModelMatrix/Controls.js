import { useState } from 'react'
import PropTypes from 'prop-types'

import { colors, spacing, typography } from '../Styles'

import { parseValue } from '../FilterRange/helpers'

import RadioGroup from '../Radio/Group'

import InputRange, { VARIANTS } from '../Input/Range'

import { DEFAULT_MIN, DEFAULT_MAX } from './reducer'

const ModelMatrixControls = ({ isNormalized, dispatch }) => {
  const domain = [DEFAULT_MIN, DEFAULT_MAX]

  const [rangeValues, setRangeValues] = useState([DEFAULT_MIN, DEFAULT_MAX])
  const [inputMin, setInputMin] = useState(rangeValues[0])
  const [inputMax, setInputMax] = useState(rangeValues[1])

  const saveMinValue = ({ value }) => {
    const newMin = parseValue({ value })

    if (newMin === rangeValues[0]) return

    if (newMin < domain[0] || newMin > rangeValues[1]) {
      setInputMin(rangeValues[0])
      return
    }

    setInputMin(newMin)

    setRangeValues([newMin, rangeValues[1]])

    dispatch({ minScore: newMin })
  }

  const saveMaxValue = ({ value }) => {
    const newMax = parseValue({ value })

    if (newMax === rangeValues[1]) return

    if (newMax < rangeValues[0] || newMax > domain[1]) {
      setInputMax(rangeValues[1])
      return
    }

    setInputMax(newMax)

    setRangeValues([rangeValues[0], newMax])

    dispatch({ maxScore: newMax })
  }

  return (
    <form
      method="post"
      onSubmit={(event) => event.preventDefault()}
      css={{
        display: 'flex',
        paddingLeft: spacing.spacious,
      }}
    >
      <RadioGroup
        legend="View"
        options={[
          {
            value: 'normalized',
            label: 'Normalized',
            legend: '',
            initialValue: isNormalized,
          },
          {
            value: 'absolute',
            label: 'Absolute',
            legend: '',
            initialValue: !isNormalized,
          },
        ]}
        onClick={({ value }) =>
          dispatch({ isNormalized: value === 'normalized' })
        }
      />

      <div
        css={{
          color: colors.structure.zinc,
          display: 'flex',
          alignItems: 'center',
          paddingLeft: spacing.spacious,
        }}
      >
        <span
          css={{
            color: colors.structure.white,
            fontWeight: typography.weight.bold,
            paddingRight: spacing.normal,
          }}
        >
          Confidence:
        </span>
        <div css={{ paddingRight: spacing.comfy }}>
          <InputRange
            label="MIN"
            value={inputMin}
            onChange={({ target: { value } }) => setInputMin(value)}
            onKeyPress={({ target: { value }, key }) => {
              if (key !== 'Enter') return
              saveMinValue({ value })
            }}
            onBlur={({ target: { value } }) => {
              saveMinValue({ value })
            }}
            variant={VARIANTS.SECONDARY}
            style={{ label: { paddingRight: spacing.comfy } }}
          />
        </div>
        <InputRange
          label="MAX"
          value={inputMax}
          onChange={({ target: { value } }) => setInputMax(value)}
          onKeyPress={({ target: { value }, key }) => {
            if (key !== 'Enter') return
            saveMaxValue({ value })
          }}
          onBlur={({ target: { value } }) => {
            saveMaxValue({ value })
          }}
          variant={VARIANTS.SECONDARY}
        />
      </div>
    </form>
  )
}

ModelMatrixControls.propTypes = {
  isNormalized: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
}

export default ModelMatrixControls
