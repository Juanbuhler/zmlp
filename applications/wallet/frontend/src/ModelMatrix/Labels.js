import PropTypes from 'prop-types'

import { constants, spacing } from '../Styles'

import { useScroller } from '../Scroll/helpers'

const ModelMatrixLabels = ({ matrix, settings: { cellDimension } }) => {
  const rowRef = useScroller({
    namespace: 'ModelMatrixHorizontal',
    isWheelEmitter: true,
    isWheelListener: true,
  })

  return (
    <div
      ref={rowRef}
      css={{
        flex: 1,
        display: 'flex',
        overflow: 'hidden',
        borderBottom: constants.borders.regular.coal,
      }}
    >
      {matrix.labels.map((label) => (
        <div
          key={label}
          css={{
            minWidth: cellDimension,
            maxWidth: cellDimension,
            padding: spacing.normal,
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            whiteSpace: 'nowrap',
            borderRight: constants.borders.regular.coal,
          }}
        >
          {label}
        </div>
      ))}
    </div>
  )
}

ModelMatrixLabels.propTypes = {
  matrix: PropTypes.shape({
    name: PropTypes.string.isRequired,
    overallAccuracy: PropTypes.number.isRequired,
    labels: PropTypes.arrayOf(PropTypes.string).isRequired,
    matrix: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
  }).isRequired,
  settings: PropTypes.shape({
    cellDimension: PropTypes.number.isRequired,
  }).isRequired,
}

export default ModelMatrixLabels
