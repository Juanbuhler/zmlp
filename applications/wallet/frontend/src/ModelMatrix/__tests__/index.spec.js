import TestRenderer, { act } from 'react-test-renderer'

import matrix from '../__mocks__/matrix'

import mockUser from '../../User/__mocks__/user'

import User from '../../User'

import { MIN_WIDTH as PANEL_MIN_WIDTH } from '../../Panel'

import ModelMatrix from '..'

jest.mock('react-tippy', () => ({
  Tooltip: jest.fn(({ children }) => <div>{children}</div>),
}))

const PROJECT_ID = '76917058-b147-4556-987a-0a0f11e46d9b'
const MODEL_ID = '621bf775-89d9-1244-9596-d6df43f1ede5'

const noop = () => () => {}

describe('<ModelMatrix />', () => {
  it('should render properly', () => {
    require('next/router').__setUseRouter({
      pathname: '/[projectId]/models/[modelId]/matrix',
      query: {
        projectId: PROJECT_ID,
        modelId: MODEL_ID,
      },
    })

    require('swr').__setMockUseSWRResponse({ data: matrix })

    const component = TestRenderer.create(
      <User initialUser={mockUser}>
        <ModelMatrix />
      </User>,
    )

    expect(component.toJSON()).toMatchSnapshot()

    // Hide Minimap
    act(() => {
      component.root.findByProps({ 'aria-label': 'Mini map' }).props.onClick()
    })

    // Open panel
    act(() => {
      component.root.findByProps({ 'aria-label': 'Preview' }).props.onClick()
    })

    // Does nothing since zoom = 1 = min
    act(() => {
      component.root.findByProps({ 'aria-label': 'Zoom Out' }).props.onClick()
    })

    // Zoom 2x
    act(() => {
      component.root.findByProps({ 'aria-label': 'Zoom In' }).props.onClick()
    })

    // Select a cell
    act(() => {
      component.root.findAllByProps({ type: 'button' })[0].props.onClick()
    })

    expect(component.toJSON()).toMatchSnapshot()

    // Back to zoom 1x
    act(() => {
      component.root.findByProps({ 'aria-label': 'Zoom Out' }).props.onClick()
    })

    // Change view to Absolute
    act(() => {
      component.root
        .findByProps({ type: 'radio', value: 'absolute' })
        .props.onClick()
    })

    act(() => {
      component.root.findByType('form').props.onSubmit({ preventDefault: noop })
    })

    // eslint-disable-next-line no-proto
    const spy = jest.spyOn(localStorage.__proto__, 'setItem')

    act(() => {
      component.root
        .findByProps({ 'aria-label': 'View Filter Panel' })
        .props.onClick({ defaultPrevented: noop })
    })

    expect(spy).toHaveBeenCalledWith(
      'rightOpeningPanelSettings',
      JSON.stringify({
        isOpen: true,
        openPanel: 'filters',
        originSize: PANEL_MIN_WIDTH,
        size: PANEL_MIN_WIDTH,
      }),
    )

    // Select a different cell
    act(() => {
      component.root.findAllByProps({ type: 'button' })[1].props.onClick()
    })

    // Deselect a cell
    act(() => {
      component.root.findAllByProps({ type: 'button' })[1].props.onClick()
    })
  })
})
