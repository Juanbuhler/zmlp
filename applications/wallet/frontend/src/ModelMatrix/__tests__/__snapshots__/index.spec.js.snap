// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModelMatrix /> should render properly 1`] = `
.emotion-0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  margin-left: -32px;
  margin-right: -32px;
  margin-bottom: -32px;
  padding-top: 1px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 4px;
  background-color: #191919;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  color: #b3b3b3;
  font-size: 17px;
  line-height: 20px;
}

.emotion-2:hover,
.emotion-2.focus-visible:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #ffffff;
}

.emotion-3 {
  margin-right: 8px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <a
      className="emotion-2"
      href="/76917058-b147-4556-987a-0a0f11e46d9b/models/621bf775-89d9-1244-9596-d6df43f1ede5"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        className="emotion-3"
        data-file-name="back.svg"
        height={20}
      />
      Model Details: 
      console_face_recognition
    </a>
  </div>
  <ModelMatrixLayout
    modelId="621bf775-89d9-1244-9596-d6df43f1ede5"
    projectId="76917058-b147-4556-987a-0a0f11e46d9b"
  />
</div>
`;
